Firestore/core/test/unit/bundle/bundle_reader_test.cc

```diff
@@ -0,0 +1,533 @@
+    }
+
+    EXPECT_EQ(reader.BytesRead(), actual_metadata.total_bytes());
+
+    return result;
+  }
+
+  void VerifyNamedQueryEncodesToOriginal(const NamedQuery& actual_read,
+                                         const ProtoNamedQuery& original) {
+    EXPECT_EQ(actual_read.ElementType(), BundleElementType::NamedQuery);
+    auto actual_proto = local_serializer.EncodeNamedQuery(actual_read);
+    nanopb::ByteString bytes = nanopb::MakeByteString(actual_proto);
+    EXPECT_TRUE(
+        msg_diff_.Compare(ProtobufParse<ProtoNamedQuery>(bytes), original))
+        << message_differences;
+    message_differences.clear();
+  }
+
+  void VerifyDocumentEncodesToOriginal(const BundleDocument& actual_read,
+                                       const ProtoDocument& original) {
+    EXPECT_EQ(actual_read.ElementType(), BundleElementType::Document);
+    auto actual_proto =
+        local_serializer.EncodeMaybeDocument(actual_read.document());
+    nanopb::ByteString bytes = nanopb::MakeByteString(actual_proto);
+    ProtoMaybeDocument maybe_document;
+    *maybe_document.mutable_document() = original;
+    EXPECT_TRUE(msg_diff_.Compare(ProtobufParse<ProtoMaybeDocument>(bytes),
+                                  maybe_document))
+        << message_differences;
+    message_differences.clear();
+  }
+
+  static void VerifyDocumentMetadataEquals(
+      const BundledDocumentMetadata& actual_read,
+      const ProtoBundledDocumentMetadata& original) {
+    EXPECT_EQ(actual_read.ElementType(), BundleElementType::DocumentMetadata);
+    EXPECT_EQ(FullPath(actual_read.key().ToString()), original.name());
+    EXPECT_EQ(actual_read.read_time(),
+              model::SnapshotVersion(Timestamp(original.read_time().seconds(),
+                                               original.read_time().nanos())));
+    EXPECT_EQ(actual_read.exists(), original.exists());
+    EXPECT_EQ(actual_read.queries(),
+              std::vector<std::string>(original.queries().begin(),
+                                       original.queries().end()));
+  }
+
+  remote::Serializer remote_serializer;
+  local::LocalSerializer local_serializer;
+  bundle::BundleSerializer bundle_serializer;
+
+ protected:
+  MessageDifferencer msg_diff_;
+  std::string message_differences;
+
+ private:
+  std::vector<std::string> elements_;
+};
+
+TEST_F(BundleReaderTest, ReadsQueryAndDocument) {
+  AddNamedQuery(LimitQuery());
+  AddNamedQuery(LimitToLastQuery());
+  AddDocumentMetadata(DocumentMetadata1());
+  AddDocument(Document1());
+
+  const auto& bundle =
+      BuildBundle("bundle-1", testutil::Version(6000004000), 1);
+  BundleReader reader(bundle_serializer,
+                      absl::make_unique<std::istringstream>(bundle));
+
+  std::vector<std::unique_ptr<BundleElement>> elements =
+      VerifyFullBundleParsed(reader, "bundle-1", testutil::Version(6000004000));
+
+  EXPECT_EQ(elements.size(), 4);
+  VerifyNamedQueryEncodesToOriginal(
```

Comment:
Note: because the `Verify` functions contain various `EXPECT` macros, it would probably help to add [`SCOPED_TRACE`](https://github.com/google/googletest/blob/master/docs/advanced.md#adding-traces-to-assertions) macros to each test that calls them.

Reply:
Done.